### PR TITLE
[WIP] Delete project resources if the corresponding namespace is missing

### LINKED ISSUE #409: Namespace deletions may lead to projects without corresponding namespaces
Sometimes the deletion of a namespace that a project references goes unnoticed in our project controller. We have to add adapt our code / add a garbage collection loop that can collect and delete such orphaned projects.

### PR DESCRIPTION
**Pull request by rfranzke**
**What this PR does / why we need it**: The project controller does now delete stale `Project` resources (i.e., those without a corresponding `Namespace` object).

**Which issue(s) this PR fixes**:
Fixes #409 

**Release note**:

```improvement user
If a `Project` resource exists without the corresponding `Namespace` then Gardener will delete it.
```

### CODE REVIEW

**Review by mvladev — CHANGES_REQUESTED** [review]
This breaks the entire contract - the idea for the `Project` resource is to control the `Namespaces` and `(Cluster)RoleBindings`.

It's like when a `ReplicaSet` is removed, the deployment controller to remove a `Deployment`

**Review by timuthy — APPROVED** [review]

**Comment by rfranzke** [comment]
There is an admission controller which is responsible for creating the namespace on project creation. If a project exists without the namespace then the namespace was deleted at the same time the project got created (race condition, see #409). In the normal case (no race) the namespace deletion is prevent by the webhook.

**Comment by mvladev** [comment]
> There is an admission controller which is responsible for creating the namespace on project creation. If a project exists without the namespace then the namespace was deleted at the same time the project got created (race condition, see #409). In the normal case (no race) the namespace deletion is prevent by the webhook.

The project [admission controller](https://github.com/gardener/gardener/blob/21c17b3de227ae6e00c5caa5b58a1bf85d1b90aa/plugin/pkg/project/namespace/admission.go#L170-L177) should not create namespaces in the first place - this is what the project controller in the gardener controller manager (should be) is responsible for.

**Comment by mvladev** [comment]
To put it simply - in an admission controller, only the resource which is being admitted SHOULD be modified (if mutation is required). No side-effects.

Otherwise if down the admission chain another admission controller rejects this resource it'll leave the system in inconsistent state.

File: pkg/controller/project/project_control.go [hunk 1]
```diff
@@ -215,8 +215,17 @@ func (c *defaultControl) ReconcileProject(obj *gardenv1beta1.Project) error {
 		return nil
 	}
 
+	// If the namespace does no longer exist we delete the project.
+	namespaceObj, err := c.k8sGardenClient.GetNamespace(*namespace)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return c.k8sGardenClient.GardenClientset().GardenV1beta1().Projects().Delete(project.Name, &metav1.DeleteOptions{})
```
Comment (adracus) [inline pkg/controller/project/project_control.go:222]: Don't we need the deletion confirmation annotation here?

**Comment by rfranzke** [comment]
FYI: Postponed until a sync with the dashboard team next week about the `Project` resources in general.

**Comment by rfranzke** [comment]
Closed in favour of #480.

**Pull request closed without merging**